frontend/src/packages/editor/operationManager.tsx

```diff
@@ -30,4 +30,98 @@ export const slateToCmsOperation = (editorContent: BlockData, operation: BaseOpe
+    {
+        Path: op.path,
+        OperationType: "insert",
+        IsNoOp: false,
+        Operation: {
+            $type: "stringOperation",
+            stringOperation: {
+                rangeStart: op.offset,
+                rangeEnd: op.offset + op.text.length,
+                newValue: op.text
+            }
+        }
+    }
+);
+
+
+const convertTextRemovalOp = (editorContent: BlockData, op: RemoveTextOperation): CMSOperation => (
+    {
+        Path: op.path,
+        OperationType: "delete",
+        IsNoOp: false,
+        Operation: {
+            $type: "stringOperation",
+            stringOperation: {
+                rangeStart: op.offset - op.text.length,
+                rangeEnd: op.offset,
+                newValue: op.text,
+            }
+        }
+    }
+);
+
+
+const normalizeElementPath = (contentBlock: CMSEditorNode, op: NodeOperation | TextOperation): number[] => {
+    switch (true) {
+        case IsCustomElement(contentBlock): return normalizeCustomElementPath(contentBlock as CustomElement, op);
+        case IsCustomTextBlock(contentBlock): return normalizeCustomTextPath(contentBlock as CustomText, op);
+    }
+
+    return [];
+}
+
+const normalizeCustomElementPath = (contentBlock: CustomElement, op: NodeOperation | TextOperation): number[] => {
+    return [] 
+}
+
+const normalizeCustomTextPath = (contentBlock: CustomText, op: NodeOperation | TextOperation): number[] => {
+    // TODO: For Mae :P - So the issue here is that fields on javascript objects are unordered, this is unlike Go where (typically) the order in which
+    // u lay them out is their actual order (assuming the compiler doesn't perform any struct packing optimizations, which it doesn't look like it does: https://github.com/golang/go/issues/10014)
+    // so this means at runtime we know the order of fields in our Go struct but not our TS struct, thus when we construct the JSON ast from the TS JSON data we are guaranteed that it conforms to the order
+    // in which they appear in Go structs.
+    
+    // The issue however is that when we get a slate operation we need to (magically) map that to a numerical value indicating the position in the Go struct/ast, for now we will just maintain a hard coded association
+    // that needs to be synchronised with: https://github.com/csesoc/website/tree/main/backend/editor/OT/data/datamodels/cmsmodel but in the future we really should try and come up with a better solution
+    // perhaps thats ur first task as backend lead :P
+    // direct mapping of: https://github.com/csesoc/website/blob/main/backend/editor/OT/data/datamodels/cmsmodel/paragraph.go#L18
+    const fieldIndexes = {
+        "text": 0,
+        "link": 1,
+        "bold": 2,
+        "italic": 3,
+        "underline": 4
+    }
+
+    const isTextOp = op.type === "insert_text" || op.type === "remove_text";
+    const finalIndex = isTextOp
+                        ? fieldIndexes.text
+                        :
```

Comment:
❓

Reply:
Michael is a draft PT to showcase an idea 😭